test/src/bgw/timer_mock.c

```diff
@@ -0,0 +1,57 @@
+#include <postgres.h>
+#include <access/xact.h>
+#include <catalog/namespace.h>
+#include <utils/builtins.h>
+#include <utils/lsyscache.h>
+#include <access/relscan.h>
+#include <access/xact.h>
+#include <storage/lmgr.h>
+#include <storage/bufmgr.h>
+#include <utils/rel.h>
+#include <utils/tqual.h>
+
+#include "timer_mock.h"
+#include "log.h"
+#include "scanner.h"
+#include "catalog.h"
+#include "compat.h"
+#include "params.h"
+
+
+static BackgroundWorkerHandle *bgw_handle = NULL;
+
+
+static bool mock_wait(TimestampTz until);
+static TimestampTz mock_current_time(void);
+
+const Timer mock_timer = {
+	.get_current_timestamp = mock_current_time,
+	.wait = mock_wait,
+};
+
+void
+timer_mock_register_bgw_handle(BackgroundWorkerHandle *handle)
+{
+	elog(WARNING, "Mock has registered new background worker");
+	bgw_handle = handle;
+}
+
+static bool
+mock_wait(TimestampTz until)
+{
+	elog(WARNING, "Running mock wait until %ld, started at %ld", until, params_get()->current_time);
+
+	if (!params_get()->mock_wait_returns_immediately && bgw_handle != NULL)
+	{
+		WaitForBackgroundWorkerShutdown(bgw_handle);
+		bgw_handle = NULL;
+	}
+	params_set_time(until);
```

Comment:
I think we need a `params_set_time_on_less_than` to allow background workers to wait for a timeout, otherwise we can get the timetravel if a process waiting for a lower `until` races one with a greater `until`. Until we implement something like that we need a warning on this function, and `mock_timer` in the header like

We need a comment on this function like

```C
/* WARNING: mock_wait must _only_ be called from the bgw_scheduler, calling it from a worker will clobber the timer state */
```